dhcpv4/async/client.go

```diff
@@ -0,0 +1,206 @@
+package async
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/fanliao/go-promise"
+	"github.com/insomniacslk/dhcp/dhcpv4"
+)
+
+// Default ports
+const (
+	DefaultServerPort = 67
+	DefaultClientPort = 68
+)
+
+// Client implements an asynchronous DHCPv4 client
+// It doesn't use the broadcast socket! Which means it should be used only when
+// the network is already established.
+type Client struct {
+	ReadTimeout  time.Duration
+	WriteTimeout time.Duration
+	LocalAddr    net.Addr
+	RemoteAddr   net.Addr
+	IgnoreErrors bool
+
+	connection   *net.UDPConn
+	cancel       context.CancelFunc
+	stopping     *sync.WaitGroup
+	receiveQueue chan *dhcpv4.DHCPv4
+	sendQueue    chan *dhcpv4.DHCPv4
+	packetsLock  sync.Mutex
+	packets      map[uint32]*promise.Promise
+	errors       chan error
+}
+
+// NewClient creates an asynchronous client
+func NewClient() *Client {
+	return &Client{
+		ReadTimeout:  dhcpv4.DefaultReadTimeout,
+		WriteTimeout: dhcpv4.DefaultWriteTimeout,
+	}
+}
+
+// Open starts the client
+func (c *Client) Open(bufferSize int) error {
+	var (
+		addr *net.UDPAddr
+		ok   bool
+		err  error
+	)
+
+	if addr, ok = c.LocalAddr.(*net.UDPAddr); !ok {
+		return fmt.Errorf("Invalid local address: %v not a net.UDPAddr", c.LocalAddr)
+	}
+
+	// prepare the socket to listen on for replies
+	c.connection, err = net.ListenUDP("udp4", addr)
+	if err != nil {
+		return err
+	}
+	c.stopping = new(sync.WaitGroup)
+	c.sendQueue = make(chan *dhcpv4.DHCPv4, bufferSize)
+	c.receiveQueue = make(chan *dhcpv4.DHCPv4, bufferSize)
+	c.packets = make(map[uint32]*promise.Promise)
+	c.packetsLock = sync.Mutex{}
+	c.errors = make(chan error)
+
+	var ctx context.Context
+	ctx, c.cancel = context.WithCancel(context.Background())
+	go c.receiverLoop(ctx)
+	go c.senderLoop(ctx)
+
+	return nil
+}
+
+// Close stops the client
+func (c *Client) Close() {
+	// Wait for sender and receiver loops
+	c.stopping.Add(2)
+	c.cancel()
+	c.stopping.Wait()
+
+	close(c.sendQueue)
+	close(c.receiveQueue)
+	close(c.errors)
+
+	c.connection.Close()
+}
+
+// Errors returns a channel where runtime errors are posted
+func (c *Client) Errors() <-chan error {
```

Comment:
would it make sense to associate errors to transaction IDs? If so, fee free to make a separate PR or open an issue for that

Reply:
The errors associated with transaction IDs are not posted to this channel, they are `Rejected` in a promise for a given request. These are the errors we were unable to associate with any of the known promises.